metagpt/tools/libs/software_development.py

```diff
@@ -0,0 +1,268 @@
+
+        >>> # Add user requirements to the exists project:
+        >>> from metagpt.tools.libs.software_development import write_prd
+        >>> project_path = '/path/to/exists_project_path'
+        >>> await write_prd("Create a new feature for the application", project_path=project_path)
+        >>> print(project_path)
+        '/path/to/exists_project_path'
+    """
+    from metagpt.actions import UserRequirement
+    from metagpt.context import Context
+    from metagpt.roles import ProductManager
+
+    ctx = Context()
+    if project_path:
+        ctx.config.project_path = Path(project_path)
+        ctx.config.inc = True
+    role = ProductManager(context=ctx)
+    msg = await role.run(with_message=Message(content=idea, cause_by=UserRequirement))
+    await role.run(with_message=msg)
+    return ctx.repo.workdir
+
+
+@register_tool(tags=["software company", "Architect"])
+async def write_design(project_path: str | Path) -> Path:
+    """Writes a design to the project repository, based on the PRD of the project.
+
+    Args:
+        project_path (str|Path): The path to the project repository.
+
+    Returns:
+        Path: The path to the project directory
+
+    Example:
+        >>> from metagpt.tools.libs.software_development import write_design
+        >>> project_path = '/path/to/project_path' # Returned by `write_prd`
+        >>> project_path = await write_desgin(project_path)
+        >>> print(project_path)
+        '/path/to/project_path'
+
+    """
+    from metagpt.actions import WritePRD
+    from metagpt.context import Context
+    from metagpt.roles import Architect
+
+    ctx = Context()
+    ctx.set_repo_dir(project_path)
+
+    role = Architect(context=ctx)
+    await role.run(with_message=Message(content="", cause_by=WritePRD))
+    return project_path
+
+
+@register_tool(tags=["software company", "Architect"])
+async def write_project_plan(project_path: str | Path) -> Path:
+    """Writes a project plan to the project repository, based on the design of the project.
+
+    Args:
+        project_path (str|Path): The path to the project repository.
+
+    Returns:
+        Path: The path to the project directory
+
+    Example:
+        >>> from metagpt.tools.libs.software_development import write_project_plan
+        >>> project_path = '/path/to/project_path' # Returned by `write_prd`
+        >>> project_path = await write_project_plan(project_path)
+        >>> print(project_path)
+        '/path/to/project_path'
+
+    """
+    from metagpt.actions import WriteDesign
+    from metagpt.context import Context
+    from metagpt.roles import ProjectManager
+
+    ctx = Context()
+    ctx.set_repo_dir(project_path)
+
+    role = ProjectManager(context=ctx)
+    await role.run(with_message=Message(content="", cause_by=WriteDesign))
+    return project_path
+
+
+@register_tool(tags=["software company", "Engineer"])
+async def write_codes(project_path: str | Path, inc: bool = False) -> Path:
```

Comment:
A clear upstream file name is required, and a clear return file name is required. The current input and output granularity is too large

Reply:
fixed. input task_path and return src path